tests/_utils/websocket.js

```diff
@@ -0,0 +1,153 @@
+                if (isJsonObject(event.data)) {
+                    window.messages.unshift(event.data);
+                }
+
+                if (window.messages.length > this.BUFFER_LENGTH) {
+                    // eslint-disable-next-line no-unused-vars
+                    window.messages = window.messages.pop();
+                }
+            };
+
+            // WebSocket "onclose" handler
+            const closeHandler = (event) => {
+                // remove event listeners
+                instance.removeEventListener('open', openHandler);
+                instance.removeEventListener('message', messageHandler);
+                instance.removeEventListener('close', closeHandler);
+            };
+
+            // add event listeners
+            instance.addEventListener('open', openHandler);
+            instance.addEventListener('message', messageHandler);
+            instance.addEventListener('close', closeHandler);
+
+            // proxy the WebSocket.send() function
+            const sendProxy = new Proxy(instance.send, {
+                apply(target, thisArg, args) {
+                    console.log('Send', args);
+                    target.apply(thisArg, args);
+                },
+            });
+
+            // replace the native send function with the proxy
+            instance.send = sendProxy;
+
+            instance.BUFFER_LENGTH = BUFFER_LENGTH;
+            // return the WebSocket instance
+            return instance;
+        },
+    });
+
+// replace the native WebSocket with the proxy
+    window.WebSocket = WebSocketProxy;
+    window.messages = [];
+}
+
+const waitForWSSubset = async (page, subset, options = {timeout: 6000}) => {
+    try {
+        return await promiseTimeout(options.timeout, checkForMessage(page, subset, {
+            subset: true,
+        }));
+    } catch (err) {
+        const messages = await page.evaluate('window.messages');
+        console.dir(messages);
+        return null;
+    }
+};
+
+// eslint-disable-next-line consistent-return
+const waitForWSMessage = async (page, message_type, {timeout = 5000}) => {
+    try {
+        return await promiseTimeout(timeout, checkForMessage(page, message_type, {
+            subset: false,
+        }));
+    } catch (err) {
+        console.log(`Could not read message ${  message_type  } in ${  timeout  }ms`);
+    }
+}
+
+const promiseTimeout = (ms, promise) => {
+    // Create a promise that rejects in <ms> milliseconds
+    const timeout = new Promise((resolve, reject) => {
+        const id = setTimeout(() => {
+            clearTimeout(id);
+            reject(new Error(`Timed out in ${ms}ms.`));
+        }, ms)
+    })
+
+    // Returns a race between our timeout and the passed in promise
+    return Promise.race([
+        promise,
+        timeout,
+    ])
+};
+
+const checkForMessage =  (page, payload, options = {
+    subset: false,
+}) => new Promise((resolve, reject) => {
+    const id = setInterval(async () => {
```

Comment:
Instead of using `setInterval`, can we create a promise for msg_type/subset and then in messageHandler we can check if there is any promise for that received msg_type/subset and resolve it? 
(It's like what we have in `deriv-api`)

Reply:
I didn't quite get what you mean here. Would you please elaborate?

Reply:
Yep, I meant that. Is it possible to use custom events to communicate so that we can resolve as soon the message we are waiting for is received